nixos/modules/services/matrix/mautrix-meta.nix

```diff
@@ -0,0 +1,210 @@
+          Configuration options should match those described in
+          [example-config.yaml](https://github.com/mautrix/meta/blob/main/example-config.yaml).
+
+          Secret tokens should be specified using {option}`environmentFile`
+          instead
+        '';
+      };
+
+      environmentFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = mdDoc ''
+          File containing environment variables to substitute when copying the configuration
+          out of Nix store to the `services.mautrix-meta.dataDir`.
+
+          Can be used for storing the secrets without making them available in the Nix store.
+
+          For example, you can set `services.mautrix-meta.settings.appservice.as_token = "$MAUTRIX_META_APPSERVICE_AS_TOKEN"`
+          and then specify `MAUTRIX_META_APPSERVICE_AS_TOKEN="{token}"` in the environment file.
+          This value will get substituted into the configuration file as as token.
+        '';
+      };
+
+      serviceDependencies = mkOption {
+        type = with types; listOf str;
+        default = optional config.services.matrix-synapse.enable config.services.matrix-synapse.serviceUnit;
+        defaultText = literalExpression ''
+          optional config.services.matrix-synapse.enable config.services.matrix-synapse.serviceUnit
+        '';
+        description = lib.mdDoc ''
+          List of Systemd services to require and wait for when starting the application service.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.mautrix-meta = {
+      description = "Mautrix-Meta, a Matrix <-> Facebook or Matrix <-> Instagram hybrid puppeting/relaybot bridge";
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ] ++ cfg.serviceDependencies;
+      after = [ "network-online.target" ] ++ cfg.serviceDependencies;
+
+      environment.HOME = fullDataDir;
+
+      preStart = ''
+        # substitute the settings file by environment variables
+        # in this case read from EnvironmentFile
+        test -f '${settingsFile}' && rm -f '${settingsFile}'
+        old_umask=$(umask)
+        umask 0177
+        ${pkgs.envsubst}/bin/envsubst \
+          -o '${settingsFile}' \
+          -i '${settingsFileUnsubstituted}'
+        umask $old_umask
+
+        # generate the appservice's registration file if absent
+        if [ ! -f '${registrationFile}' ]; then
+          ${pkg}/bin/mautrix-meta \
+            --generate-registration \
+            --config='${settingsFile}' \
+            --registration='${registrationFile}'
+        fi
+
+        umask 0177
+        ${pkgs.yq}/bin/yq -s '.[0].appservice.as_token = .[1].as_token
+          | .[0].appservice.hs_token = .[1].hs_token
+          | .[0]' '${settingsFile}' '${registrationFile}' \
+          > '${settingsFile}.tmp'
+        mv '${settingsFile}.tmp' '${settingsFile}'
+        umask $old_umask
```

Comment:
I think this is backwards. It copies from the registration file which is randomly generated to the settings file. This makes the configured value for the settings useless. (And contradicts the docs for `environmentFile` which imply that you can use it to configure these tokens)

Reply:
I dont think this is true completely. The values are generated randomly only if the value is not configured in the config. If it is, the registration file will contain these. So its possible to first put in config values, then let it generate, and it will contain the ones you set, not random ones.

Reply:
I guess that is true but at least after the first generation they values will be ignored. So this results in pinning the values from the first configuration which is surprising.

Reply:
I will solve this issue by regenerating the registration file if as_token and hs_token are present in the configuration.

Reply:
I've changed the behavior so that the registration file gets as_token and hs_token replaced with the ones from settings, if they are set. And if not, the values from registration are used to allow random generation by the bridge as well.

Reply:
I'm also still not sold on avoiding regenerating the registration. By externally generating the registration file I have already avoided lots of bugs because the registration is stale.

It seems that there are only a few bits of state we actually need:

- `hs_token`
- `as_token`
- `sender_localpart`

`hs_token` and `as_token` should be copied into the main config anyways. So I wonder what re-generating the registration does. Maybe it just copies them so that isn't an issue?

`sender_localpart` appears to be randomly generated but doesn't need to be. I just have it fixed to `mautrix-meta-facebook`. I don't think it would be an issue to do that. However the registration generation step doesn't pull this from the config so if we want to fix it we would need to update or generate the registration outself.

The advantage is that we can now update the values that should be updated. Notably:

- `id`. Doesn't really matter but is nice.
- `url` Should always be up-to-date.
- `rate_limited`, `push_ephemeral`, ... These are all configuration for the homeserver that should be updated as the bridge takes advantage of new changes.
- `users`. Should be updated if the bridge user templates are updated. But this is a pretty disruptive change so shouldn't be a common change past initial generation.

Basically this file mixes state and configuration which is unfortunate. The current approach is probably fine as I don't think anything major should change once you have the bridge configured. However it does cause a lot of confusion while you are configuring the bridge as some settings get stuck and it isn't clear why. There is also minor long-term maintenance with various flags but that is much less of an issue.

Just for the record this is what my registration looks like right now:

```nix
registrationData = {
	as_token = secrets.as_token;
	hs_token = secrets.hs_token;

	id = cfg-settings.appservice.id;
	url = cfg-settings.appservice.address;
	sender_localpart = cfg-settings.appservice.id;

	namespaces = {
		users = [{
			exclusive = true;
			regex = "^${lib.escapeRegex "@${cfg-settings.appservice.bot.username}:${cfg-settings.homeserver.domain}"}$";
		} {
			exclusive = true;
			# Note, this isn't fully reusable but could probably be made by properly massaging `bridge.username_template`.
			regex = "^@facebook_.*:${lib.escapeRegex cfg-settings.homeserver.domain}$";
		}];
		aliases = [];
	};

	# Ideally these would come from the bridge binary, but updating them manually isn't very difficult and can allow working around bugs more easily.
	rate_limited = false;
	"de.sorunome.msc2409.push_ephemeral" = true;
	push_ephemeral = true;
};
```

Of course I think this isn't the interesting part, the interesting part is how to deploy this without putting `as_token` and `hs_token` into the Nix store.

Given the current state of things I think my preferred solution would be something like this:

1. If registration exists and `hs_token` or `as_token` aren't set substitute from the registration into the config.
2. Regenerate the registration.
3. Update `hs_token`, `as_token`, `id` and `sender_localpart` from the config.

This way the registration is always up to date, with the only exception being if the `hs_token` and `as_token` aren't explicitly set in the config they will be auto-generated and preserved.

A cleaner approach may be to completely separate `hs_token` and `as_token` into a separate secrets file, then just substitute them into both the config and registration. This file can either be provide by the user (example via `nixops`) or automatically generated on first start. Then the actual state is completely separated from the config.

But I think the current approach is probably fine.

Reply:
I have asked tulir and `hs_token`, `as_token` are indeed regenerated every time registration is generated, even if they are already in the config. It was different for some other bridges, but now it's like this. `sender_localpart` is not used at all, it's irrelevant.

I am not sure if using generate registration is a good idea then. It always changes the config which can be unexpected.

Reply:
Maybe we could use similar approach to `registrationData` from `mautrix-facebook`, but instead `envsubst` the registration file, to allow for passing the options inside the `environmentFile`.
The registration service would then generate the file completely by itself using the `envsubst`. If the environment variables were missing, it could generate random HS_TOKEN, AS_TOKEN. (first run, later runs would have to keep it)

What do you think about this?

Reply:
> I am not sure if using generate registration is a good idea then. It always changes the config which can be unexpected.

What do you mean by this? I described saving the `hs_token` and `as_token` so I don't think it should be an issue if they are regenerated in the base step, because then they will be overwritten with the saved or user-configured values.

> What do you think about this?

I'm not 100% sure what you are describing. Are you saying just replace the built-in registration with our own script? Then extract the `{as,hs}_token` values and sub them into the config? I think that would work well.

Reply:
> What do you mean by this? 
I meant the `mautrix-meta --generate-registration`. At first I thought it would be great, but since I learned that it changes the values in config, including as_token and hs_token, I began to be afraid of what it could break.

> I'm not 100% sure what you are describing. Are you saying just replace the built-in registration with our own script? Then extract the {as,hs}_token values and sub them into the config? I think that would work well.
i am suggesting doing approach like with `settingsFileUnsubstituted`, but for the `registrationFile`. And moreover, if the user does not configure the hs_token, as_token fields, they will be automatically generated - to still allow for approach where you don't want to use something like sops, and want to avoid having secrets in the store.

So there would be something like this to configure:
```
      registrationData = {
        id = cfg.settings.appservice.id;
        as_token = "$AS_TOKEN";
        hs_token = "$HS_TOKEN";

        namespaces = {
          users = [
            {
              exclusive = true;
              regex = escapeRegex "@${cfg.settings.appservice.bot_username}:${cfg.settings.homeserver.domain}";
            }
            {
              exclusive = true;
              regex = "@${puppetRegex}:${escapeRegex cfg.settings.homeserver.domain}";
            }
          ];
          aliases = [];
        };

        url = cfg.settings.appservice.address;
        sender_localpart = "mautrix-meta-sender";

        rate_limited = false;
        "de.sorunome.msc2409.push_ephemeral" = true;
        push_ephemeral = true;
      };
    };
```
Then it would be written into nix store by `pkgs.formats.yaml.generate`, and lastly, the registration service would `envsubst` this file into `/var/lib/mautrix-meta-X/meta-registration.yaml`. And in case AS_TOKEN, HS_TOKEN were missing, they would be generated. I am thinking of saving these into `/var/lib/mautrix-meta-X/secrets`, since both config and registration can get replaced and this seems like the easiest solution to make sure of not losing them.

I think that instead of using `yq`, we could stick to just `envsubst` even in the config.

Reply:
Personally I would rather just overwrite the `{hs,as}_token` in the file generated by mautrix-meta with jq/yq as it is more robust. We know exactly what keys should have which values so it seems easiest to use these tools that make the exact syntax-aware substitution rather than some text substitution.

If anything I would prefer to do everything this way. Instead of `environmentFile` I'd rather have something like `secretSettings` which is a JSON file that gets merged in. It is just a more precise operation. Especially if we just need these two tokens.

Reply:
I've updated it. What do you think about it now?

Reply:
I won't be able to look again until tomorrow. I'll get back then.

Reply:
This looks good.